[wasm] Initial test support on windows

radekdoulik opened the pull request:
Add support to run libraries tests on windows. Examples of how to run the tests:

    .\build.cmd libs.tests -test -os browser
    .\dotnet.cmd build /t:Test /p:TargetOS=Browser src\libraries\System.Collections.Concurrent\tests
    .\dotnet.cmd build /t:Test /p:TargetOS=Browser /p:JSEngine="SpiderMonkey" src\libraries\System.Text.Json\tests

radekdoulik committed:
[wasm] Initial test support on windows

radekdoulik committed:
Fix dotnet call and %RSP_FILE% addition

Doh, need to call dotnet.exe with `.exe` extension in `.cmd` file.

Fix the condition so that we don't add `$RSP_FILE` as well on windows.

radekdoulik committed:
Add target to run libraries tests

radekdoulik committed:
Do not run on jsc

It is failing on Windows, opened
https://github.com/dotnet/runtime/issues/49884 to not forget about it.

radekdoulik committed:
Few more changes for Wasm.Build.Tests

AaronRobinsonMSFT commented on src/mono/wasm/wasm.proj:
```diff
@@ -230,4 +230,28 @@
+      </Libraries>
+    </ItemGroup>
+
+    <Exec Command="dotnet build %(Libraries.RootDir)%(Libraries.Directory) /p:JSEngine=%(Libraries.Engine) /t:Test /p:TargetOS=Browser /p:TargetArchitecture=wasm" />
```
This seems like it is only for non-DOM scenarios. Are we also going to be enabling DOM scenarios? The Crypto APIs will not be able to run on non-DOM environments. I assume this is complex given the need for some browser driver being installed.

See https://github.com/dotnet/runtime/pull/49511

radekdoulik replied:
Yes, this is now only for stand alone engines. I plan to add target to run the tests in browser on windows in future PR.

radekdoulik committed:
Add documentation on how to run libraries tests

on windows

radical commented:
We should enable tests on windows here, which would validate this PR too.

radekdoulik committed:
Merge remote-tracking branch 'origin/main' into pr-wasm-windows-tests

radekdoulik committed:
Enable running tests

radekdoulik committed:
Add manifest entry for mono-aot-cross.exe

radekdoulik committed:
Copy icu libs to runtime pack

radekdoulik committed:
Use the provisioned EmSDK when available

radekdoulik committed:
Set version info for mono sgen .exe binaries

radekdoulik committed:
Set FallbackFileVersion for mono-aot-cross.exe

radekdoulik committed:
Only warn about stale open files

On CI we are getting:

    D:\workspace\_work\1\s\eng\testing\tests.targets(49,5): error MSB3231: Unable to remove directory "D:\workspace\_work\1\s\artifacts\bin\WebAssembly.Browser.AOT.Test\net6.0-Release\browser-wasm\publish\". The process cannot access the file 'Microsoft.DotNet.XHarness.TestRunners.Xunit.dll' because it is being used by another process. [D:\workspace\_work\1\s\src\tests\FunctionalTests\WebAssembly\Browser\AOT\WebAssembly.Browser.AOT.Test.csproj]

`Microsoft.DotNet.XHarness.TestRunners.Xunit` is used as package
reference by few projects, so it is probably still opened by one of the
build processes.

radekdoulik committed:
Try stop build processes to see if it helps

with:

    error MSB3231: Unable to remove directory "D:\workspace\_work\1\s\artifacts\bin\WebAssembly.Browser.AOT.Test\net6.0-Release\browser-wasm\publish\". The process cannot access the file 'Microsoft.DotNet.XHarness.TestRunners.Xunit.dll' because it is being used by another process.

radekdoulik commented:
/azp run

radekdoulik commented:
/azp run runtime-staging

radekdoulik commented:
/azp run [runtime-staging]

radekdoulik committed:
Rebuild

radekdoulik committed:
Fix indentation

radekdoulik committed:
Use powershell task

radekdoulik committed:
Leave just wasm/windows in the platform matrix

radekdoulik committed:
Do not remove publish dir for now

radekdoulik committed:
Set CI build timeout

radekdoulik committed:
Changes to make "Send to helix" step start

radekdoulik committed:
Set configuration when running the tests

radekdoulik committed:
Add smoke test

radekdoulik committed:
Remove extra step to run on helix

radekdoulik commented:
/azp run

radekdoulik committed:
Merge remote-tracking branch 'origin/main' into pr-wasm-windows-tests

radekdoulik committed:
Install npm with pip

radekdoulik committed:
Try to use nodetool task

radekdoulik committed:
Use npm exec -c ...

radekdoulik committed:
Set PATH to jsvu and check v8 before build

radekdoulik committed:
Update the targets after merge with main

radekdoulik committed:
Call dotnet tool restore before running tests

radekdoulik committed:
Update the scripts

radekdoulik committed:
Use newer nodetool

radekdoulik committed:
Specify the os for jsvu as well

radekdoulik committed:
Cleaning

radekdoulik committed:
Put delete of publish dir back for non-windows

radekdoulik committed:
Merge remote-tracking branch 'origin/main' into pr-wasm-windows-tests

radekdoulik committed:
Restore platform matrix

akoeplinger commented on src/installer/pkg/sfx/Microsoft.NETCore.App/Directory.Build.props:
```diff
@@ -171,6 +171,7 @@
     <PlatformManifestFileEntry Include="libmonosgen-2.0.dylib" IsNative="true" />
     <PlatformManifestFileEntry Include="llc" IsNative="true" />
     <PlatformManifestFileEntry Include="mono-aot-cross" IsNative="true" />
+    <PlatformManifestFileEntry Include="mono-aot-cross.exe" IsNative="true" FallbackFileVersion="6.0.0" />
```
why do we need the `FallbackFileVersion`?

radekdoulik replied:
Because the installer requires it for `.exe` files. Can't find the place in the code right now.

akoeplinger replied:
It's not set for `createdump.exe` though :)

I think this actually verifies that the .exe has a non-0.0.0.0 file version, which I think should've been fixed by your `NativeVersion.rc` change. Can you double check this still happens?

radekdoulik replied:
I think it was failing here IIRC: https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.SharedFramework.Sdk/src/GeneratePlatformManifestEntriesFromTemplate.cs#L84-L99

akoeplinger commented on src/mono/wasm/README.md:
```diff
@@ -50,25 +52,56 @@ The latest engines can be installed with jsvu (JavaScript engine Version Updater
 `npm install jsvu -g`
 
-* Run jsvu:
+* Run jsvu and install `v8`, `SpiderMonkey`,or `JavaScriptCore` engines:
```
```suggestion
* Run jsvu and install `v8`, `SpiderMonkey`, or `JavaScriptCore` engines:
```

akoeplinger commented on src/mono/wasm/README.md:
```diff
@@ -50,25 +52,56 @@ The latest engines can be installed with jsvu (JavaScript engine Version Updater
+
+`npm install jsvu -g`
+
+* Run jsvu and install `v8`, `SpiderMonkey`,or `JavaScriptCore` engines:
```
```suggestion
* Run jsvu and install `v8`, `SpiderMonkey`, or `JavaScriptCore` engines:
```

radekdoulik committed:
Do not run the smoke test, leave it for next PR

akoeplinger commented on src/mono/wasm/README.md:
```diff
@@ -50,25 +52,56 @@ The latest engines can be installed with jsvu (JavaScript engine Version Updater
 
+### Windows
+
+To run the libraries tests on windows, we have `RunTests` target. Without setting additional properties, it will run tests for all libraries on `V8` and `SpiderMonkey` engines:
```
```suggestion
To run the libraries tests on windows, we have the `RunTests` target. Without setting additional properties, it will run tests for all libraries on `v8` and `SpiderMonkey` engines:
```

akoeplinger commented on src/mono/wasm/README.md:
```diff
@@ -50,25 +52,56 @@ The latest engines can be installed with jsvu (JavaScript engine Version Updater
+`dotnet build /t:RunTests wasm.proj`
+
+* `Library` property can be used to specify which library tests to run
+* `VESngine` property can be used to specify which engine to use
```
there's a typo here compared to the wasm.proj, but I think this should actually be called `JSEngine`.

akoeplinger commented on src/mono/wasm/README.md:
```diff
@@ -50,25 +52,56 @@ The latest engines can be installed with jsvu (JavaScript engine Version Updater
+* `Library` property can be used to specify which library tests to run
+* `VESngine` property can be used to specify which engine to use
+
+For example to run tests for `System.Collections.Concurrent` on `V8` engine, use:
```
```suggestion
For example to run tests for `System.Collections.Concurrent` on the `v8` engine, use:
```

akoeplinger commented on src/mono/wasm/README.md:
```diff
@@ -50,25 +52,56 @@ The latest engines can be installed with jsvu (JavaScript engine Version Updater
+
+For example to run tests for `System.Collections.Concurrent` on `V8` engine, use:
+
+`dotnet build /t:RunTests /p:Library="System.Collections.Concurrent" /p:VSEngine="v8" wasm.proj`
```
```suggestion
`dotnet build /t:RunTests /p:Library="System.Collections.Concurrent" /p:JSEngine="v8" wasm.proj`
```

akoeplinger replied:
not sure we need this special target, the documented way is using `./build.sh libs.tests -test -os browser` to build and run all library tests: https://github.com/dotnet/runtime/blob/main/docs/workflow/testing/libraries/testing.md#testing-libraries

You could pass down the property to specify the JS engine via `/p:JSEngine=v8`

radekdoulik replied:
I added the target to have something similar to https://github.com/dotnet/runtime/blob/main/src/mono/wasm/Makefile#L179-L187

akoeplinger replied:
yeah like I said that can be achieved by `build.cmd libs.tests -test -os browser /p:JSEngine=v8`, which is consistent with how other parts of the repo do it. I'd rather not add another custom target.

radekdoulik replied:
I think you are right. The individual library run has similar verbosity to running it directly too. I have updated the README and removed the target.

akoeplinger commented on eng/testing/WasmRunnerTemplate.cmd:
```diff
@@ -0,0 +1,45 @@
+set SCENARIO=%3
+
+if [%HELIX_WORKITEM_UPLOAD_ROOT%] == [] (
+	set XHARNESS_OUT="%EXECUTION_DIR%xharness-output"
```
nit: I think most of the other Windows .cmd files uses spaces for indentation instead of tabs

radekdoulik committed:
Update src/mono/wasm/README.md

Co-authored-by: Alexander Köplinger <alex.koeplinger@outlook.com>

radekdoulik committed:
Update src/mono/wasm/README.md

Co-authored-by: Alexander Köplinger <alex.koeplinger@outlook.com>

radekdoulik committed:
Update src/mono/wasm/README.md

Co-authored-by: Alexander Köplinger <alex.koeplinger@outlook.com>

radekdoulik committed:
Update src/mono/wasm/README.md

Co-authored-by: Alexander Köplinger <alex.koeplinger@outlook.com>

radekdoulik committed:
Update src/mono/wasm/README.md

Co-authored-by: Alexander Köplinger <alex.koeplinger@outlook.com>

radekdoulik committed:
Fix typo, s/VSEngine/JSEngine

radekdoulik committed:
Fix one more typo

radekdoulik committed:
Use spaces for indentation

radekdoulik committed:
Remove the RunTests target and update README

akoeplinger commented on eng/pipelines/runtime-staging.yml:
```diff
@@ -284,12 +284,14 @@ jobs:
   parameters:
```
nit, to be fixed in a follow-up PR: please add the purpose of this node as a comment like the others do:

```
#
# Builds Browser_wasm on Windows
#
- template: /eng/pipelines/common/platform-matrix.yml
   ...
```

akoeplinger commented on eng/testing/tests.targets:
```diff
@@ -45,15 +46,16 @@
     <!-- delete the BundleDir and PublishDir in CI builds to save disk space on build agents since they're no longer needed -->
-    <RemoveDir Condition="'$(ContinuousIntegrationBuild)' == 'true'" Directories="$(BundleDir);$(PublishDir)" />
+    <RemoveDir Condition="'$(ContinuousIntegrationBuild)' == 'true'" Directories="$(BundleDir)" ContinueOnError="WarnAndContinue" />
+    <RemoveDir Condition="'$(ContinuousIntegrationBuild)' == 'true' and '$(OS)' != 'Windows_NT'" Directories="$(PublishDir)" ContinueOnError="WarnAndContinue" />
```
removing the PublishDir should work on Windows too right?

akoeplinger commented on src/installer/pkg/sfx/Microsoft.NETCore.App/Directory.Build.props:
```diff
@@ -171,6 +171,7 @@
     <PlatformManifestFileEntry Include="libmonosgen-2.0.dylib" IsNative="true" />
     <PlatformManifestFileEntry Include="llc" IsNative="true" />
     <PlatformManifestFileEntry Include="mono-aot-cross" IsNative="true" />
+    <PlatformManifestFileEntry Include="mono-aot-cross.exe" IsNative="true" FallbackFileVersion="6.0.0" />
```
we should follow-up to see why the file version doesn't get set in the .exe. Hardcoding the version here will just bite us once the repo moves to 7.0.0

radekdoulik replied:
The file version is set by this change here https://github.com/dotnet/runtime/pull/49830/commits/4ca80bbd69c4ac30cc7a9a2062e898d941d75048. IIRC it is mandatory to have the fallback version. As we are not going to ship windows build, it hopefully doesn't matter which version we set as the fallback version.

I will try to build it without the fallback version set to see whether it will still fail the build. It is possible that I was only getting failures before adding the version to the `.exe` file.

akoeplinger replied:
@radekdoulik did you have a chance to try a build without the `FallbackFileVersion`?

radekdoulik replied:
Yes, looks like it is not needed anymore. https://github.com/dotnet/runtime/pull/51229

Let see if it still builds with the restored platform matrix.

akoeplinger reviewed (approved):
Looks good with a few comments that we need to follow-up, but I don't think they need to hold this PR :)
